lib/Arch/Sleigh/Thumb2Arch.cpp

```diff
@@ -24,54 +25,78 @@
 #include <remill/OS/OS.h>
 
 #include "Arch.h"
+#include "Thumb.h"
 
 namespace remill {
 namespace sleighthumb2 {
 
+
 //ARM7_le.sla"
-class SleighThumb2Arch final : public remill::sleigh::SleighArch,
-                               public remill::AArch32ArchBase {
+SleighThumb2Decoder::SleighThumb2Decoder(const remill::Arch &arch)
+    : SleighDecoder(arch, "ARM7_le.sla", "ARMtTHUMB.pspec",
+                    {{"ISAModeSwitch", std::string(kThumbModeRegName)}}) {}
+
+
+void SleighThumb2Decoder::InitializeSleighContext(
+    remill::sleigh::SingleInstructionSleighContext &ctxt) const {
+  ctxt.GetContext().setVariableDefault("TMode", 1);
+}
+
+//TODO(Ian): this has code duplication with SleighX86Arch couldnt come up with a way to share implementation and not run into more
+// annoying virtual inheretance from remill Arch. If we go back to virtual Arch then maybe we could just add another virtual inheratance of
+// Arch. All of these are bad tho.
+class SleighThumbArch : public AArch32ArchBase {
  public:
-  SleighThumb2Arch(llvm::LLVMContext *context_, OSName os_name_,
-                   ArchName arch_name_)
+  SleighThumbArch(llvm::LLVMContext *context_, OSName os_name_,
+                  ArchName arch_name_)
       : ArchBase(context_, os_name_, arch_name_),
-        SleighArch(context_, os_name_, arch_name_, "ARM7_le.sla",
-                   "ARMtTHUMB.pspec"),
-        AArch32ArchBase(context_, os_name_, arch_name_) {}
+        AArch32ArchBase(context_, os_name_, arch_name_),
 
+        decoder(*this) {}
 
-  uint64_t MaxInstructionSize(bool permit_fuse_idioms) const final {
-    return 4;
+  virtual DecodingContext CreateInitialContext(void) const override {
+    return DecodingContext().PutContextReg(std::string(kThumbModeRegName), 1);
   }
 
-  uint64_t MinInstructionSize(void) const final {
-    return 2;
+  virtual OperandLifter::OpLifterPtr
+  DefaultLifter(const remill::IntrinsicTable &intrinsics) const override {
+    return this->decoder.GetOpLifter();
+  }
+
+  virtual bool DecodeInstruction(uint64_t address, std::string_view instr_bytes,
+                                 Instruction &inst,
+                                 DecodingContext context) const override {
+    return decoder.DecodeInstruction(address, instr_bytes, inst, context);
   }
 
-  uint64_t MinInstructionAlign(void) const final {
+
+  // TODO(pag): Eventually handle Thumb2 and unaligned addresses.
```

Comment:
@2over12 Do we handle unaligned addresses now?

Reply:
Not really, are instructions at misaligned addresses a thing, afaik instructions always align them by clearing the last bit regardless of if it mode switches or not

Reply:
That was my understanding as well